Custom nuget-feed no longer used, breaking package updates

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package ecosystem

nuget

### Package manager version

_No response_

### Language version

_No response_

### Manifest location and content before the Dependabot update

Our repos have multiple csproj files (under various subfolders like src/ and test/) and Directory.Build.props at the root


### dependabot.yml content

```
version: 2
registries:
  my-artifactory-nuget:
    type: nuget-feed
    url: https://my-artifactory-url
    username: username
    password: password

updates:
- package-ecosystem: nuget
  directory: /
  schedule:
    interval: daily
    time: "21:00"
    timezone: America/New_York
  allow:
  - dependency-name: MyPackage.Name
  registries:
  - my-artifactory-nuget
```

### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

On Feb 4th we starting seeing dependabot failures reporting "Dependabot couldn't find any dependency files in the directory"

Reading the logs however, you can see that dependabot is no longer accessing our custom nuget feed, and instead using nuget.org. This obviously doesn't work for us as our internal packages are only on our internal feed. 

Eventually in the logs you see errors like 

```
error NU1101: Unable to find package <MyPackage>. No packages exist with this id in source(s): nuget.org
```

### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

<img width="708" alt="Image" src="https://github.com/user-attachments/assets/22617ff5-0337-414b-83ae-12964b20508d" />

### Smallest manifest that reproduces the issue

_No response_

Chiming in to say we're also experiencing it. Our packages are hosted at `nuget.pkg.github.com` so I'm presuming this affects any projects with a custom feed.

Excerpt of our logs, if it helps

```plaintext
2025/02/13 11:06:45 INFO Temporarily removing `global.json` from `/home/dependabot/dependabot-updater/repo`.
2025/02/13 11:06:45 INFO Restoring `global.json` to `/home/dependabot/dependabot-updater/repo`.
2025/02/13 11:06:45 INFO Discovering build files in workspace [/home/dependabot/dependabot-updater/repo].
2025/02/13 11:06:45 INFO   No dotnet-tools.json file found.
2025/02/13 11:06:45 INFO   Discovered [global.json] file.
2025/02/13 11:06:45 INFO   Discovering projects beneath [.].
2025/02/13 11:06:46 INFO   No packages.config file found.
2025/02/13 11:06:46 INFO InstallDotnetSdks == true; retaining `global.json` contents.
2025/02/13 11:06:47 INFO InstallDotnetSdks == true; retaining `global.json` contents.
2025/02/13 11:07:03 WARN   Error determining dependencies from `/home/dependabot/dependabot-updater/repo/MyProject.Tests/MyProject.Tests.csproj`:
STDOUT:
  Determining projects to restore...
/home/dependabot/dependabot-updater/repo/MyProject.Tests/MyProject.Tests.csproj : error NU1101: Unable to find package MyPackage.Shared.Common. No packages exist with this id in source(s): nuget.org
/home/dependabot/dependabot-updater/repo/MyProject/MyProject.csproj : error NU1101: Unable to find package MyPackage.Shared.Common. No packages exist with this id in source(s): nuget.org [/home/dependabot/dependabot-updater/repo/MyProject.Tests/MyProject.Tests.csproj]
  Failed to restore /home/dependabot/dependabot-updater/repo/MyProject/MyProject.csproj (in 14.02 sec).
  Failed to restore /home/dependabot/dependabot-updater/repo/MyProject.Tests/MyProject.Tests.csproj (in 14.02 sec).


STDERR:


2025/02/13 11:07:04 INFO   No packages.config file found.
2025/02/13 11:07:04 INFO InstallDotnetSdks == true; retaining `global.json` contents.
2025/02/13 11:07:05 INFO InstallDotnetSdks == true; retaining `global.json` contents.
2025/02/13 11:07:07 WARN   Error determining dependencies from `/home/dependabot/dependabot-updater/repo/MyProject/MyProject.csproj`:
STDOUT:
  Determining projects to restore...
/home/dependabot/dependabot-updater/repo/MyProject/MyProject.csproj : error NU1101: Unable to find package MyPackage.Shared.Common. No packages exist with this id in source(s): nuget.org
  Failed to restore /home/dependabot/dependabot-updater/repo/MyProject/MyProject.csproj (in 1.04 sec).
```

--- a/nuget/lib/dependabot/nuget/file_fetcher.rb
+++ b/nuget/lib/dependabot/nuget/file_fetcher.rb
@@ -26,6 +26,7 @@ module Dependabot
 
       sig { override.returns(T::Array[DependencyFile]) }
       def fetch_files
+        NativeHelpers.normalize_file_names
         NativeHelpers.install_dotnet_sdks
         discovery_json_reader = DiscoveryJsonReader.run_discovery_in_directory(
           repo_contents_path: T.must(repo_contents_path),

--- a/nuget/lib/dependabot/nuget/native_helpers.rb
+++ b/nuget/lib/dependabot/nuget/native_helpers.rb
@@ -287,6 +287,22 @@ module Dependabot
       end
 
       sig { void }
+      def self.normalize_file_names
+        # environment variables are required and the following will generate an actionable error message if they're not
+        _dependabot_repo_contents_path = ENV.fetch("DEPENDABOT_REPO_CONTENTS_PATH")
+
+        # this environment variable is directly used
+        dependabot_home = ENV.fetch("DEPENDABOT_HOME")
+
+        command = [
+          "pwsh",
+          "#{dependabot_home}/dependabot-updater/bin/normalize-file-names.ps1"
+        ].join(" ")
+        output = SharedHelpers.run_shell_command(command)
+        puts output
+      end
+
+      sig { void }
       def self.install_dotnet_sdks
         return unless Dependabot::Experiments.enabled?(:nuget_install_dotnet_sdks)
 

--- a/nuget/spec/dependabot/nuget/file_fetcher_spec.rb
+++ b/nuget/spec/dependabot/nuget/file_fetcher_spec.rb
@@ -72,6 +72,8 @@ RSpec.describe Dependabot::Nuget::FileFetcher do
           discovery_json_content = discovery_content_hash.to_json
           File.write(discovery_json_path, discovery_json_content)
         end
+      allow(Dependabot::Nuget::NativeHelpers)
+        .to receive(:normalize_file_names)
       # stub call to `fetch_file_from_host` because it expects an empty directory and other things that make the test
       # more difficult than it needs to be
       allow(file_fetcher_instance)

--- a/nuget/updater/common.ps1
+++ b/nuget/updater/common.ps1
@@ -134,3 +134,20 @@ function Install-TargetingPacks([string]$sdkInstallDir, [string[]]$targetingPack
         Remove-Item -Path $archiveName
     }
 }
+
+function Repair-FileCasingForName([string]$fileName) {
+    # Get-ChildItem is case-insensitive
+    $discoveredFiles = Get-ChildItem $env:DEPENDABOT_REPO_CONTENTS_PATH -r -inc $fileName
+    foreach ($file in $discoveredFiles) {
+        # `-cne` = Case-sensitive Not Equal
+        if ($file.Name -cne $fileName) {
+            $newName = "$($file.Directory)/$fileName"
+            Write-Host "Renaming '$file' to '$newName'"
+            Rename-Item -Path $file -NewName $newName
+        }
+    }
+}
+
+function Repair-FileCasing() {
+    Repair-FileCasingForName -fileName "NuGet.Config"
+}

--- a/nuget/updater/main.ps1
+++ b/nuget/updater/main.ps1
@@ -21,6 +21,7 @@ function Get-Files {
         --api-url $env:DEPENDABOT_API_URL `
         --job-id $env:DEPENDABOT_JOB_ID
     $script:operationExitCode = $LASTEXITCODE
+    Repair-FileCasing
 }
 
 function Update-Files {

--- a/nuget/updater/normalize-file-names.ps1
+++ b/nuget/updater/normalize-file-names.ps1
@@ -1,0 +1,14 @@
+Set-StrictMode -version 2.0
+$ErrorActionPreference = "Stop"
+
+. $PSScriptRoot\common.ps1
+
+try {
+    Repair-FileCasing
+}
+catch {
+    Write-Host $_
+    Write-Host $_.Exception
+    Write-Host $_.ScriptStackTrace
+    exit 1
+}

